### PR TITLE
disttask: fix wrong plan when filterByRole

### LINKED ISSUE #48368: add index and import into may generate wrong plan with more nodes
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
1. deploy 2 tidb with label "background" and "".
2. set global tidb_enable_dist_task=on;
3. add index

### 2. What did you expect to see? (Required)
Add index run on tidb with label “background" with subtask cnt cnt1.

### 3. What did you see instead (Required)
Add index run on tidb with label “background" with subtask cnt cnt2 = 2*cnt1.

### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: close #48368

Problem Summary:
1. serverNodes used in OnNextStageBatch didn't filter by role.
2. serverNodes may change during planning and dispatching phase, may lead to inconsistency of task nodes.
### What is changed and how it works?
```
server:= d.GetEligibleInstance()
server:=d.filterByRole(server)
subtasks := d.OnNextStageBatch(server)
d.dispatchSubtasks(subtasks, server)
```
Using the above logic, the planner will use consistent serverNodes in plan and dispatching phase.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Enhance planning of dist task
```

### CODE REVIEW

**Review by D3Hunter — APPROVED** [review]

File: pkg/disttask/framework/dispatcher/dispatcher.go [hunk 1]
```diff
@@ -545,7 +552,26 @@ func (d *BaseDispatcher) onNextStage() (err error) {
...
-		metas, err := d.OnNextSubtasksBatch(d.ctx, d, d.Task, nextStep)
+		/// select all available TiDB nodes for task.
+		serverNodes, filter, err := d.GetEligibleInstances(d.ctx, d.Task)
+		logutil.Logger(d.logCtx).Debug("eligible instances", zap.Int("num", len(serverNodes)))
+
+		if err != nil {
+			return err
+		}
+		if filter {
+			/// filter by role.
```
Comment (tangenta) [inline pkg/disttask/framework/dispatcher/dispatcher.go:563]:
```suggestion
```

The comment did not provide more information :)

Comment (ywqzzy) [reply pkg/disttask/framework/dispatcher/dispatcher.go:563]: lol

**Review by tangenta — APPROVED** [review]

**Pull request merged**